Programs/brlapi_client.c

```diff
@@ -1602,10 +1606,29 @@ int BRLAPI_STDCALL brlapi__enterTtyModeWithPath(brlapi_handle_t *handle, int *tt
 
   /* OK, Now we know where we are, so get the effective control of the terminal! */
   ttytreepath = getenv("WINDOWPATH");
-  if (!ttytreepath && getenv("DISPLAY"))
+  if (!ttytreepath && (getenv("DISPLAY") || getenv("WAYLAND_DISPLAY"))) {
     /* Cope with some DMs which are not setting WINDOWPATH (e.g. gdm 3.12) */
     ttytreepath = getenv("XDG_VTNR");
 
+#ifdef HAVE_SD_SESSION_GET_VT
+    if (!ttytreepath) {
+      /* Fallback to asking logind */
+      char *id;
+      unsigned vtnr;
+      int ret = sd_pid_get_session(0, &id);
+      if (ret == 0 && id != NULL) {
```

Comment:
Unfortunately, this will not be enough in all cases. If `sd_pid_get_session` returns an error, then we need to start guessing the session. Which requires looping over all sessions and finding the "best" graphical one.

I can cook up a patch if that is helpful.

Reply:
I don't think we want to guess, that most often provides more harm than good :/

Reply:
Yup, unfortunately we really have to. There is no way around it and it is safe because users can only have one graphical session by definition. Some people have not accepted that one yet though.

The thing is that the processes of the graphical session end up being started by systemd. And then they technically don't actually belong to any user session. There are some further long term ideas and more details … but right now, guessing is definitely the right thing to do!

Reply:
Note that you *should* use `sd_pid_get_session` if it returns something sane. Only when that fails fall back to guessing.

Reply:
> unfortunately we really have to

Why? In which situation sd_pid_get_session might not find it?

> users can only have one graphical session by definition

?

one can run startx several times on different VTs. We do not want to mix Braille output in that case.

> The thing is that the processes of the graphical session end up being started by systemd. And then they technically don't actually belong to any user session. There are some further long term ideas and more details …

So you mean that when you start an application inside a graphical session, systemd takes over it and forgets essential informations such as where it is actually running??